### DISCUSSION TITLE
Any way to use a debugger?

### DISCUSSION THREAD

**Discussion by ror3d** [statiqdev/Discussions#119] [Q&A]
When experimenting with things I'm not sure how they really work, I find myself wishing for a way to debug the execution, to see what the inputs and outputs to everything is. I am guessing that this might not be possible from inside a razor template, but maybe there is some way to write some testing code in the C# application code that I can step into with a debugger that could give me more insight?

I am using VSCode and compiling from the powershell terminal, but I'd happily use anything else that could help with this.

**Comment by corbob** [comment]
I'm in the same boat here. I have managed to get a debugger to attach, and stop on a breakpoint... but it's not actually been helpful for me. I would greatly like to stop debugging in one of the razor templates, but that doesn't work. 

I am able to enter the debugger in vscode. The code I have that allows me to stop the debugger is:

```csharp
using Statiq.App;
using Statiq.Web;

Console.WriteLine("Breakpoint here");
await Bootstrapper // Also Breakpoint here...
	.Factory
	.CreateWeb(args)
	.RunAsync();
Console.WriteLine("Final Breakpoint here...");
```

When I run: `dotnet run -- --attach` I end up with output:

```
~/git/coryknox.dev [gitConfigPost ≡ +1 ~2 -0 !]> dotnet run -- --attach           
/Users/coryknox/git/coryknox.dev/src/Navigation.cs(5,23): warning CS8618: Non-nullable property 'Title' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/coryknox/git/coryknox.dev/coryknox.dev.csproj]
/Users/coryknox/git/coryknox.dev/src/Navigation.cs(6,23): warning CS8618: Non-nullable property 'Link' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/coryknox/git/coryknox.dev/coryknox.dev.csproj]
/Users/coryknox/git/coryknox.dev/src/Navigation.cs(7,23): warning CS8618: Non-nullable property 'ToolTip' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/coryknox/git/coryknox.dev/coryknox.dev.csproj]                                                                 
Breakpoint here                                                                                     
[INFO] Waiting for a debugger to attach to process 89446 (or press a key to continue)...
```

It then waits as it says for me to attach the debugger. Once the debugger it continues until the Final Breakpoint here... line and then enters the debugger.

The other way I've been able to debug it, and actually hit breakpoints before the bootstrapper is to use the `.NET Core Launch (console)` debug configuration that the C# extension created for me. It then stopped on the first breakpoint, but when I tried to step into the Bootstrapper it just skipped over everything with a message about "Just my code" being enabled, and none of the cshtml breakpoints I set were ever hit 😦

**Comment by daveaglick** [comment]
Note that since Statiq generators are just normal .NET projects, running in debug mode straight from Visual Studio, Rider, etc. is always possible. That's actually how I debug my generators most of the time these days.

So there are a few options:
- Run the generator application in debug mode from an IDE.
- Use the `--attach` flag as described above which will pause execution at the beginning and wait for a debugger to attach.
- Use the `--debug` flag which will pop a dialog asking you to select a debugger (in most cases...this doesn't always work depending on what's installed on the system, Windows vs. Linux, etc.).
- Write your own `Debugger.Launch();` statement somewhere in your code.

Note that Statiq also bundles debugging symbols with [SourceLink](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink) support, so stepping into Statiq source code should be possible (providing you're IDE supports SourceLink and/or decompilation and you've configured it to do so - I.e. turned off "just my code").

As for debugging Razor assemblies, I know a couple folks have had limited luck doing so but that's not "official" supported right now in the sense that I haven't looked into it very much and haven't spent time ensuring it works. Definitely something I'd like to do, just hasn't risen to top priority yet. That said, I do know the Razor engine is outputting assemblies and PDB files right now, so in theory all the parts are there to make this work.

**Comment by julia-and** [comment]
I find that Razor debugging works in Visual Studio (2022 and 2019) but not in Rider. You have to disable "Just My Code" in the debugger settings.
The experience isn't perfect but it can help.

**Reply by McNerdius** [reply]
Disabling "Just My Code" for VS Code can be done by editing your debug launch configuration(s) (`launch.json`), adding `"justMyCode":false` to applicable configs.  See here:  https://github.com/OmniSharp/omnisharp-vscode/blob/master/debugger-launchjson.md#just-my-code
